numba/tests/test_typedlist.py

```diff
@@ -689,6 +826,319 @@ def foo():
+        foo_true_expected.append(d)
+        self.assertEqual(foo_true_expected, foo_true_received)
+        self.assertEqual(foo_false_expected, foo_false_received)
+        self.assertEqual(List, type(foo_true_received))
+        self.assertEqual(list, type(foo_false_received))
+
+    def test_square_bracket_builtin_from_nested_iter(self):
+        def foo():
+            l = [[1, 2, 3], [4, 5, 6]]
+            return l
+        foo_true, foo_false = self._njit_both(foo)
+        foo_true_received, foo_false_received = foo_true(), foo_false()
+        foo_true_expected, foo_false_expected = List(), list([[1, 2, 3],
+                                                              [4, 5, 6]])
+        a = List()
+        [a.append(i) for i in (1, 2, 3)]
+        b = List()
+        [b.append(i) for i in (4, 5, 6)]
+        foo_true_expected.append(a)
+        foo_true_expected.append(b)
+
+        self.assertEqual(foo_true_expected, foo_true_received)
+        self.assertEqual(foo_false_expected, foo_false_received)
+
+        self.assertEqual(List, type(foo_true_received))
+        self.assertEqual(List, type(foo_true_received[0]))
+        self.assertEqual(List, type(foo_true_received[1]))
+
+        self.assertEqual(list, type(foo_false_received))
+        self.assertEqual(list, type(foo_false_received[0]))
+        self.assertEqual(list, type(foo_false_received[1]))
+
+    def test_square_bracket_builtin_from_iter_type_coercion(self):
+        def foo():
+            l = [1, 1.0]
+            return l
+        foo_true, foo_false = self._njit_both(foo)
+        foo_true_expected = List()
+        foo_true_expected.append(1.0)
+        foo_true_expected.append(1.0)
+        foo_false_expected = [1.0, 1.0]
+        foo_true_received, foo_false_received = foo_true(), foo_false()
+        # FIXME: the first item is coerced to float
+
+        self.assertEqual(foo_true_expected, foo_true_received)
+        self.assertEqual(foo_false_expected, foo_false_received)
+        self.assertEqual(List, type(foo_true_received))
+        self.assertEqual(list, type(foo_false_received))
+
+    def test_square_bracket_builtin_from_iter_type_exception(self):
+        @njit
+        def foo():
+            l = [1, "a"]
+            return l
+        with self.assertRaises(TypingError) as raises:
+            foo()
+        # FIXME: the error message could be more specific
+        self.assertIn(
+            "Type of variable 'l' cannot be determined",
+            str(raises.exception),
+        )
+
+
+class TestConversionListToImmutableTypedList(MemoryLeakMixin, TestCase):
+
+    def test_simple_conversion(self):
+        @njit(_disable_reflected_list=True)
+        def foo(lst):
+            return lst
+        # Python list goes in and Numba immutable typed list comes out
+        received = foo([1, 2, 3])
+        expected = List()
+        [expected.append(i) for i in (1, 2, 3)]
+        # FIXME: this may fail if mutability is included in equality
```

Comment:
```suggestion
        # NOTE: this may fail if mutability is included in equality
```
Unless there's something to actually fix?

Reply:
Yeah, I am not sure about this one and what the semantics should be.